pkg/controller/job/job_controller.go

```diff
@@ -1554,13 +1565,18 @@ func getBackoff(queue workqueue.RateLimitingInterface, key interface{}) time.Dur
 	return calculated
 }
 
-// countValidPodsWithFilter returns number of valid pods that pass the filter.
+// getValidPodsWithFilter returns the valid pods that pass the filter.
 // Pods are valid if they have a finalizer and, for Indexed Jobs, a valid
 // completion index.
-func countValidPodsWithFilter(job *batch.Job, pods []*v1.Pod, uncounted sets.String, expectedRmFinalizers sets.String, filter func(*v1.Pod) bool) int {
-	result := len(uncounted)
+func getValidPodsWithFilter(job *batch.Job, pods []*v1.Pod, uncounted sets.String, expectedRmFinalizers sets.String, filter func(*v1.Pod) bool) []*v1.Pod {
+	result := []*v1.Pod{}
 	for _, p := range pods {
 		uid := string(p.UID)
+
+		if uncounted.Has(uid) {
```

Comment:
Not sure if this is a bug or not, but: A pod first is marked as Failed, so we increment the backoff. Then the pod enters the uncounted set, at which point we might count it again.

Unless I missed something.

Reply:
The changes in this PR does not change the original logic of counting new succeeded or new failed pods. The original version is [here](https://github.com/kubernetes/kubernetes/blob/487c44323922fa9a0d217ac3a70eb5943a121fbb/pkg/controller/job/job_controller.go#L1560).

In [line 1561](https://github.com/kubernetes/kubernetes/blob/487c44323922fa9a0d217ac3a70eb5943a121fbb/pkg/controller/job/job_controller.go#L1561), result is initialised to `len(uncounted)`. The `if` block in the current PR `if uncounted.Has(uid) {` does exactly that. I think if this changes here double count the backoff, it should lead to double counting of `status.succeess` and `status.failure` in the master as well. Please let me know if this makes sense.

Reply:
> The changes in this PR does not change the original logic of counting new succeeded or new failed pods.

That's where I think there is a problem. In a first sync, a pod that failed is added to `result`, then we remove the finalizer and add it to the uncounted set. In a second sync, the pod is in the uncounted set, so it's added to `result` again. Isn't this pod counted twice as a new failure across two syncs?

Reply:
I went through the code and recorded my observations below.

When the pod fails, the following events happen in the controller loop:

1. `E1` => Compute newly failed pods. To be considered for newly failed pods, at least one of the following conditions should be satisfied.
    1. Pod should be in the failed state
    1. Pod should be in the uncounted set 
1. `E2` => Newly failed pods whose finalisers are not removed are added to uncounted pods and the job status gets updated
2. `E3` => Pod finalisers are removed
3. `E4` => The uncounted pods are removed and the job status gets updated
4. `E5` => The back-off gets updated

`E4` failure is the only event that can result in uncounted pods in the job status. When `E4` fails, `E5` does not get executed; hence, the back-off is not updated.

Let me know if this makes sense.

Reply:
I made some changes to address the issue. I have not squashed the commit so that you can have a look at the diff. With this new change, for back-off calculation, only newly succeeded/failed pods are included; uncounted pods are not counted. Also, changed the location where update back-off is called; now, it's called when the status update is called to add uncounted pods. Please let me know if this change makes sense.

Reply:
why add these pods to the list if you are filtering them out in newBackoffRecord?

Reply:
The reason why I have not removed in `getValidPodsWithFilter` is to keep it consistent with the current implementation as in [here](https://github.com/kubernetes/kubernetes/blob/487c44323922fa9a0d217ac3a70eb5943a121fbb/pkg/controller/job/job_controller.go#L1560). In the master, we are explicitly setting `result := len(uncounted)` that includes uncounted in the result. I did not want to change the behavior as it might break something. WDYT? I can try that out tomorrow if that is ok!

Reply:
But then you can just add it to the counters, instead of adding it unnecessarily to the slices.

Reply:
Done